src/diff/mod.rs

```diff
@@ -1,36 +1,285 @@
+#![allow(dead_code, unused_variables)]
+
 use crate::file_system::{Directory, Label};
 use nonempty::NonEmpty;
+use crate::file_system::{DirectoryContents, File};
+use std::cmp::Ordering;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+#[derive(Debug)]
+pub struct DiffError {
+    reason: String
+}
+
+impl From<String> for DiffError {
+    fn from(reason: String) -> Self {
+        DiffError { reason }
+    }
+}
+
+pub struct Diff {
+    pub created: Vec<CreateFile>,
+    pub deleted: Vec<DeleteFile>,
+    pub moved: Vec<MoveFile>,
+    pub modified: Vec<ModifiedFile>,
+}
 
-pub struct Location {}
-
-pub enum Change {
-    AddLine {
-        path: NonEmpty<Label>,
-        location: Location,
-        contents: String,
-    },
-    RemoveLine {
-        path: NonEmpty<Label>,
-        location: Location,
-    },
-    MoveFile {
-        old_path: NonEmpty<Label>,
-        new_path: NonEmpty<Label>,
-    },
-    CreateFile {
-        path: NonEmpty<Label>,
-    },
-    DeleteFile {
-        path: NonEmpty<Label>,
-    },
-}
-
-pub struct Diff(pub Vec<Change>);
+pub struct CreateFile {
+    pub path: NonEmpty<Label>,
```

Comment:
Actually, why aren't these `Path`s?

Reply:
Switched to using Path and simplified some other things along the way

Reply:
Seems like by doing so I screwed up the order of path elements (they are reversed now). Need to look into the API closer..

Reply:
`Path`s are as you would read them. `foo/bar/baz.rs` is `(foo, vec![bar, baz.rs])`